hamilton/driver.py

```diff
@@ -900,6 +901,30 @@ def visualize_execution(
             show_schema=show_schema,
         )
 
+    @capture_function_usage
+    def export_json(
+        self,
+        final_vars: List[str],
+        inputs: Dict[str, Any] = None,
+        overrides: Dict[str, Any] = None,
+    ) -> str:
+        """Method to create JSON representation of the Graph.
+
+        :param final_vars: The final variables to compute.
+        :param inputs: Optional. The inputs to the DAG.
+        :param overrides: Optional. Overrides to the DAG.
+        :return: JSON string representation of the graph.
+        """
+        nodes, user_nodes = self.graph.get_upstream_nodes(final_vars, inputs, overrides)
+        Driver.validate_inputs(self.graph, self.adapter, user_nodes, inputs, nodes)
+        all_nodes = nodes | user_nodes
+
+        # Create JSON representation of the graph
+        # TODO: Do we also want to include information on whether the node is IS_USER_INPUT and IS_OUTPUT
+        node_dicts = [node.as_dict() for node in all_nodes]
+        sorted_nodes = sorted(node_dicts, key=lambda x: x["name"])
+        return json.dumps(sorted_nodes)
+
```

Comment:
So we want this to:

1. Use https://github.com/DAGWorks-Inc/hamilton/blob/main/hamilton/graph_types.py#L17
2. Add the as_dict() function to HamiltonNode
3. change this function to mention it is exporting only the what would be executed given the final vars, inputs, and overrides. Maybe this means the name should incorporate `execution` in it to mirror the other functions in the driver?

@elijahbenizzy anything to add/disagree with?

Reply:
@skrawcz mostly. I think that it could be one step cleaner -- we have `HamiltonGraph`, which has a `from_graph` function. We should have an `as_dict` function there that delegates to the one in `HamiltonNode`. Then this just has to call `from_graph` on `self.graph` and it should be good!

Reply:
From graph wouldn't help here because the graph is being filtered down though?

Reply:
Yes, my thoughts were this method would be similar to the `visualize_execution` method. It accepts a list of `final_vars` and then constructs a JSON representation of the graph that you would see with the `visualize_execution` method, rather than a JSON representation of the entire `FunctionGraph`.

On adding the `as_dict()` method to the `HamiltonNode`. There are some things omitted from the HamiltonNode that are in the `Node` that I was originally exporting. Specfically:
 - [source](https://github.com/DAGWorks-Inc/hamilton/pull/635/files#diff-6b4a15b6f5263150ab8674431da707b5ea1c4a855e07a42ffb79ff5aa56cea86R224) - This is the source code of the `callable` function. I notice the `callable` isn't in the the `HamiltonNode`, is this by design?  There is the `originating_functions`, However I could not work out what these functions are, I think I need to spend longer reading the source code.
 - [depended_on_by](https://github.com/DAGWorks-Inc/hamilton/pull/635/files#diff-6b4a15b6f5263150ab8674431da707b5ea1c4a855e07a42ffb79ff5aa56cea86R223) - This is also not in the HamiltonNode.

On the name, I agree it is a bit confusing at the minute. Some other suggestions (from chatgpt):
 - export_execution
 - serialize_execution
 - jsonify_execution_path

Reply:
> There is the originating_functions,

Node can have a bunch of wrapped functions. Originating functions is what decorators add to. So the source code you're after will be in the first entry that you'd want (IIRC).

> [depended_on_by](https://github.com/DAGWorks-Inc/hamilton/pull/635/files#diff-6b4a15b6f5263150ab8674431da707b5ea1c4a855e07a42ffb79ff5aa56cea86R223) - This is also not in the HamiltonNode.

That's true, though it's not required to recreate the graph. dependend_on_by is a mirror of dependencies, I just decided to have two to know which way you're traversing the graph. So you could reconstruct this part yourself.

Reply:
Ahh yep, see your point. Then yeah, I'm a fan of `export_execution`, and using the nodes... 

To add to what @skrawcz said, you can get the following:

1. `source` -- from `inspect.getsourcelines(originating_functions[0]) if originating_functions else None` (or something like that).  This has to do with the way hamilton works - the `callable` does not necessarily correspond to the actual function. Instead, it is sometimes wrapped with some special logic. The originating_functions are the one you want, as they're guarenteed to be the ones you defined (if they exist, there are some edge-cases of dynamically generated nodes). IIRC it is always the first. In 99% of the cases there is just one, so use the first in the list.
2. `depended_on_by` -- I didn't want to include this as it encodes the same information twice -- the fact that we included it is a little odd TBH (although it does make traversing it easier).

I think this should give you enough to work through it?